### PR TITLE
reject build requests for binary builds if not providing binary inputs

### PR DESCRIPTION
**Pull request by bparees**
new output:
```
$ oc start-build somebinarybuildcfg
Error from server: can't instantiate binary BuildConfig p1/somebinarybuildcfg: must provide a build input via --from-dir|--from-file|--from-repo
```

fixes https://github.com/openshift/origin/issues/9261

### CODE REVIEW

**Comment by bparees** [comment]
@gabemontero ptal.

**Comment by bparees** [comment]
[test]

**Comment by gabemontero** [comment]
Igtm (if the test failure was from the changes I missed  how the changes
would have caused the failure)

**Comment by bparees** [comment]
yeah appears to be a new flake, opened https://github.com/openshift/origin/issues/9680

**Comment by bparees** [comment]
@smarterclayton can you ptal as well?  i want to make sure i'm not missing something about the philosophy of binary builds.

File: pkg/build/generator/generator.go [hunk 1]
```diff
@@ -218,9 +218,11 @@ func (g *BuildGenerator) Instantiate(ctx kapi.Context, request *buildapi.BuildRe
 	if err != nil {
 		return nil, err
 	}
-
+	if request.Binary == nil && bc.Spec.Source.Dockerfile == nil && bc.Spec.Source.Git == nil && len(bc.Spec.Source.Images) == 0 {
```
Comment (smarterclayton) [inline pkg/build/generator/generator.go]: Why is this here and not in validation?

File: pkg/build/generator/generator.go [hunk 2]
```diff
@@ -333,6 +335,9 @@ func (g *BuildGenerator) Clone(ctx kapi.Context, request *buildapi.BuildRequest)
 		if err != nil && !errors.IsNotFound(err) {
 			return nil, err
 		}
+		if build.Spec.Source.Dockerfile == nil && build.Spec.Source.Git == nil && len(build.Spec.Source.Images) == 0 {
```
Comment (smarterclayton) [inline pkg/build/generator/generator.go]: Same question.

File: pkg/build/generator/generator.go (hunk 1 above)
Comment (bparees) [inline pkg/build/generator/generator.go]: what validation?  client-side we don't even retrieve the buildconfig that i saw, so we can't do it there.

and the buildrequest object validation also doesn't currently have a need to retrieve the buildconfig object in order to do validation, it seems burdensome to introduce that call in the validation flow.

Comment (bparees) [inline pkg/build/generator/generator.go]: (and the associated confusion of "what error do we return to a user when buildrequest validation failed because we got an spurious/temporary error retrieving the buildconfig associated with the buildrequest")

Comment (smarterclayton) [inline pkg/build/generator/generator.go]: Validate the build you are creating

Comment (bparees) [inline pkg/build/generator/generator.go]: I think my approach allows for a nicer error message  (in validation it's going to need to be a generic "everything is nil, you must specify one" message).  but there's no doubt we were missing that validation, so i'll put it there.

Comment (bparees) [inline pkg/build/generator/generator.go]: go figure, it turns out we do have that validation, but binary is not remaining nil because we copy the value from the buildconfig when creating the build.  will fix that.

**Comment by bparees** [comment]
@smarterclayton reworked, ptal.  output is:

```
$ oc start-build centos
The Build "centos-2" is invalid.
spec.source: Invalid value: {"Binary":null,"Dockerfile":null,"Git":null,"Images":null,"ContextDir":"","SourceSecret":null,"Secrets":[]}: must provide a value for at least one source input(git, binary, dockerfile, images). For binary builds, use --from-dir|--from-repo|--from-file
```

and for --from-build:

```
$ oc start-build centos --from-build=centos-4
The Build "centos-5" is invalid.
spec.source: Invalid value: {"Binary":null,"Dockerfile":null,"Git":null,"Images":null,"ContextDir":"","SourceSecret":null,"Secrets":[]}: must provide a value for at least one source input(git, binary, dockerfile, images). For binary builds, use --from-dir|--from-repo|--from-file
```

**Comment by bparees** [comment]
and yeah that means oc create operations for builds are going to see:
` must provide a value for at least one source input(git, binary, dockerfile, images). For binary builds, use --from-dir|--from-repo|--from-file`

which doesn't make a ton of sense.  which is one reason it was nicer to have an error coming out of the rest endpoint.

**Comment by bparees** [comment]
@smarterclayton bump

File: pkg/build/api/validation/validation.go [hunk 3]
```diff
@@ -142,7 +142,7 @@ func validateCommonSpec(spec *buildapi.CommonSpec, fldPath *field.Path) field.Er
 	s := spec.Strategy
 
 	if s.DockerStrategy != nil && s.JenkinsPipelineStrategy == nil && spec.Source.Git == nil && spec.Source.Binary == nil && spec.Source.Dockerfile == nil && spec.Source.Images == nil {
-		allErrs = append(allErrs, field.Invalid(fldPath.Child("source"), spec.Source, "must provide a value for at least one of source, binary, images, or dockerfile"))
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("source"), spec.Source, "must provide a value for at least one source input(git, binary, dockerfile, images). For binary builds, use --from-dir|--from-repo|--from-file"))
```
Comment (smarterclayton) [inline pkg/build/api/validation/validation.go]: We shouldn't be sending command line instructions back from the API

**Comment by bparees** [comment]
[testextended][extended:core(starting a build using CLI)]

**Comment by bparees** [comment]
@smarterclayton i am not doing checking inside the instantiate/clone code because at that point we're getting back a StatusError from the create call, which means instantiate/clone needs to:
1) cast the error to a statuserror
2) unwrap the validation error within the status error (if it's even still present in its original form, i didn't even get that far before getting frustrated w/ it being a fiddly mess)
3) check if the validation error pertains to spec.source and appears to indicate a binary build related problem
4) return a bad request error to the instantiate/clone caller

then the instantiate/clone caller (client in this case) has to interpret the bad request error again, to present a useful error message to the user.

imho the validation error being returned directly from the instantiate/clone api is sufficient for a client to do the right thing.  They tried to instantiate a build, they got a validation error.

**Comment by bparees** [comment]
@smarterclayton bump.

File: pkg/build/generator/generator.go [hunk 4]
```diff
@@ -218,9 +218,8 @@ func (g *BuildGenerator) Instantiate(ctx kapi.Context, request *buildapi.BuildRe
 	if err != nil {
 		return nil, err
 	}
-
 	if buildutil.IsPaused(bc) {
-		return nil, errors.NewInternalError(&GeneratorFatalError{fmt.Sprintf("can't instantiate from BuildConfig %s/%s: BuildConfig is paused", bc.Namespace, bc.Name)})
+		return nil, errors.NewBadRequest(fmt.Sprintf("can't instantiate from BuildConfig %s/%s: BuildConfig is paused", bc.Namespace, bc.Name))
```
Comment (smarterclayton) [inline pkg/build/generator/generator.go]: We should be setting a cause on errors like this - can you open an issue?

File: pkg/cmd/cli/cmd/startbuild.go [hunk 5]
```diff
@@ -717,3 +729,14 @@ func WaitForBuildComplete(c osclient.BuildInterface, name string) error {
 		}
 	}
 }
+
+func isInvalidSourceInputsError(err error) bool {
+	if err != nil {
+		if statusErr, ok := err.(*kerrors.StatusError); ok {
+			if statusErr.ErrStatus.Reason == "Invalid" && strings.Contains(statusErr.ErrStatus.Message, "spec.source") {
```
Comment (smarterclayton) [inline pkg/cmd/cli/cmd/startbuild.go]: I don't like this.  This is really hacky.

Comment (smarterclayton) [inline pkg/cmd/cli/cmd/startbuild.go]: You can already walk the causes for "spec.source" - do that instead.

File: pkg/build/generator/generator.go (hunk 4 above)
Comment (bparees) [inline pkg/build/generator/generator.go]: what, against upstream?

File: pkg/cmd/cli/cmd/startbuild.go [hunk 6]
```diff
@@ -717,3 +729,18 @@ func WaitForBuildComplete(c osclient.BuildInterface, name string) error {
 		}
 	}
 }
+
+func isInvalidSourceInputsError(err error) bool {
+	if err != nil {
+		if statusErr, ok := err.(*kerrors.StatusError); ok {
+			if statusErr.ErrStatus.Reason == "Invalid" {
+				for _, cause := range statusErr.ErrStatus.Details.Causes {
```
Comment (bparees) [inline pkg/cmd/cli/cmd/startbuild.go]: @smarterclayton better?  doesn't seem better.

Comment (smarterclayton) [inline pkg/cmd/cli/cmd/startbuild.go]: No, just use the source field "spec.source" instead of the message (you get that separate from message).  And don't check Reason against string "Invalid", check the error type against `api/errors.IsInvalid`

File: pkg/cmd/cli/cmd/startbuild.go [hunk 7]
```diff
@@ -717,3 +729,18 @@ func WaitForBuildComplete(c osclient.BuildInterface, name string) error {
 		}
 	}
 }
+
+func isInvalidSourceInputsError(err error) bool {
+	if err != nil {
+		if statusErr, ok := err.(*kerrors.StatusError); ok {
+			if kerrors.IsInvalid(statusErr) {
+				for _, cause := range statusErr.ErrStatus.Details.Causes {
+					if cause.Field == "spec.source" {
```
Comment (bparees) [inline pkg/cmd/cli/cmd/startbuild.go]: @smarterclayton how about now?

**Comment by smarterclayton** [comment]
LGTM

**Comment by bparees** [comment]
[merge]
